src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.MetadataDb.cs

```diff
@@ -231,7 +231,22 @@ private void Enlarge()
                 Debug.Assert(!_isLocked, "Appending to a locked database");
 
                 byte[] toReturn = _data;
-                _data = ArrayPool<byte>.Shared.Rent(toReturn.Length * 2);
+
+                // Allow the data to grow up to maximum possible capacity (~2G bytes) before encountering overflow.
+                // Note: Array.MaxLength exists only on .NET 6 or greater,
+                // so for the other versions value is hardcoded and taken from internal Array.MaxArrayLength
+#if NET6_0_OR_GREATER
+                int maxArrayLength = Array.MaxLength;
+#else
+                const int maxArrayLength = 0x7FEFFFFF;
+#endif
```

Comment:
```suggestion
                const int MaxArrayLength = 0x7FEFFFFF;
#if NET6_0_OR_GREATER
                Debug.Assert(MaxArrayLength == Array.MaxLength);
#else
#endif
```

Reply:
I have re-checked the source, and seems that there is another constant for byte arrays (`Array.MaxByteArrayLength`) which is equal to `Array.MaxLength`. I have tested and it works as it should, so I will replace it in addition to this change.